[Bug] gitea crash while squash-merging

gitea version: 1.14.0+dev-158-gc178a3643

to reproduce:
1. create new test repo
2. create master, dev branch
3. protect branches (master, dev)
4. make some pulls & merge to dev
5. squash-merge dev to master


```
[Macaron] 2020-11-09 21:22:21: Started POST /api/internal/hook/pre-receive/test/squash for [::1]
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-list --max-count=1 0e9406b8e1229168bd3bbf8d63ecc481bcb940da ^a1225a7c6b38c0da7c83a14985a8fc533e652748
[Macaron] 2020-11-09 21:22:21: Completed POST /api/internal/hook/pre-receive/test/squash 200 OK in 13.035397ms
[Macaron] 2020-11-09 21:22:22: Started POST /api/internal/hook/post-receive/test/squash for [::1]
[Macaron] 2020-11-09 21:22:22: Completed POST /api/internal/hook/post-receive/test/squash 200 OK in 8.190653ms
[Macaron] 2020-11-09 21:22:22: Completed POST /test/squash/pulls/3/merge 302 Found in 597.512363ms
[Macaron] 2020-11-09 21:22:22: Started GET /test/squash/pulls/3 for 127.0.0.1
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-list --count a1225a7c6b38c0da7c83a14985a8fc533e652748
[git-module] stdout:
2

[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= merge-base -- 0e9406b8e1229168bd3bbf8d63ecc481bcb940da refs/pull/3/head
[git-module] stdout:
0e9406b8e1229168bd3bbf8d63ecc481bcb940da

[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= log 0e9406b8e1229168bd3bbf8d63ecc481bcb940da...refs/pull/3/head --pretty=format:%H
[git-module] stdout:
30ee358d13685d64e974a3bbf8a0e6c5a379f155
04806b46b14cda20923b3d7f7279f2c3476a7a15
12b451d950d0db3b81fa0cae73046cefc01d9dcc
42405e2b47de4deb625c8fabb4f4905f90641c91
37291966d67c91c0cd1861eafd9dee4eb68f59ec
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= diff -z --name-only 0e9406b8e1229168bd3bbf8d63ecc481bcb940da...refs/pull/3/head
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-list 0e9406b8e1229168bd3bbf8d63ecc481bcb940da...a1225a7c6b38c0da7c83a14985a8fc533e652748
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= config --get commit.gpgsign
[git-module] stdout:
a1225a7c6b38c0da7c83a14985a8fc533e652748

[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-list --max-count=1 0e9406b8e1229168bd3bbf8d63ecc481bcb940da ^a1225a7c6b38c0da7c83a14985a8fc533e652748
[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-parse --verify refs/heads/main
[git-module] stdout:
a1225a7c6b38c0da7c83a14985a8fc533e652748

[git-module] /home/6543/tmp/gitea_master/repositories/test/squash.git: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= rev-list --count a1225a7c6b38c0da7c83a14985a8fc533e652748
[git-module] stdout:
2

panic: runtime error: slice bounds out of range [:603] with capacity 576

goroutine 12 [running]:
code.gitea.io/gitea/modules/references.FindAllIssueReferences(0xc0096a0240, 0x234, 0x457f29, 0x37b6c35, 0x8dd0b9)
        /home/6543/git/gitea/modules/references/references.go:269 +0x4b4
code.gitea.io/gitea/modules/repofiles.UpdateIssuesCommit(0xc00959db80, 0xc009169c20, 0xc009e862b0, 0x1, 0x1, 0xc008fca9cb, 0x4, 0xc00969edf0, 0xb)
        /home/6543/git/gitea/modules/repofiles/action.go:140 +0xf3
code.gitea.io/gitea/services/repository.commitRepoAction(0xc009169c20, 0xc0093f5b90, 0xc009e4e118, 0x1, 0x1, 0xc009611ea0, 0x0)
        /home/6543/git/gitea/services/repository/push.go:252 +0xb87
code.gitea.io/gitea/services/repository.pushUpdates(0xc009d864e0, 0x1, 0x1, 0x0, 0x0)
        /home/6543/git/gitea/services/repository/push.go:185 +0x23ca
code.gitea.io/gitea/services/repository.handle(0xc000348a00, 0x1, 0x14)
        /home/6543/git/gitea/services/repository/push.go:33 +0x87
code.gitea.io/gitea/modules/queue.(*WorkerPool).doWork(0xc001442080, 0x2cd89e0, 0xc001376600)
        /home/6543/git/gitea/modules/queue/workerpool.go:383 +0x896
code.gitea.io/gitea/modules/queue.(*WorkerPool).addWorkers.func1(0xc001442080, 0x2cd89e0, 0xc001376600)
        /home/6543/git/gitea/modules/queue/workerpool.go:238 +0x45
created by code.gitea.io/gitea/modules/queue.(*WorkerPool).addWorkers
        /home/6543/git/gitea/modules/queue/workerpool.go:237 +0x56
```

--- a/modules/references/references.go
+++ b/modules/references/references.go
@@ -235,77 +235,81 @@ func findAllIssueReferencesMarkdown(content string) []*rawReference {
 	return findAllIssueReferencesBytes(bcontent, links)
 }
 
+func convertFullHTMLReferencesToShortRefs(re *regexp.Regexp, contentBytes *[]byte) {
+	// We will iterate through the content, rewrite and simplify full references.
+	//
+	// We want to transform something like:
+	//
+	// 0         1         2         3         4         5         6
+	// 012345678901234567890123456789012345678901234567890123456789012345789
+	// this is a https://ourgitea.com/git/owner/repo/issues/123456789, foo
+	// https://ourgitea.com/git/owner/repo/pulls/123456789
+	//
+	// Into something like:
+	//
+	// this is a #123456789, foo
+	// !123456789
+
+	pos := 0
+	for {
+		// re looks for something like: (\s|^|\(|\[)https://ourgitea.com/git/(owner/repo)/(issues)/(123456789)(?:\s|$|\)|\]|[:;,.?!]\s|[:;,.?!]$)
+		match := re.FindSubmatchIndex((*contentBytes)[pos:])
+		if match == nil {
+			break
+		}
+		// match is a bunch of indices into the content from pos onwards so if our content is:
+
+		// To simplify things let's just add pos to all of the indices in match
+		for i := range match {
+			match[i] += pos
+		}
+
+		// match[0]-match[1] is whole string
+		// match[2]-match[3] is preamble
+
+		// move the position to the end of the preamble
+		pos = match[3]
+
+		// match[4]-match[5] is owner/repo
+		// now copy the owner/repo to end of the preamble
+		endPos := pos + match[5] - match[4]
+		copy((*contentBytes)[pos:endPos], (*contentBytes)[match[4]:match[5]])
+
+		// move the current position to the end of the newly copied owner/repo
+		pos = endPos
+
+		// Now set the issue/pull marker:
+		//
+		// match[6]-match[7] == 'issues'
+		(*contentBytes)[pos] = '#'
+		if string((*contentBytes)[match[6]:match[7]]) == "pulls" {
+			(*contentBytes)[pos] = '!'
+		}
+		pos++
+
+		// Then add the issue/pull number
+		//
+		// match[8]-match[9] is the number
+		endPos = pos + match[9] - match[8]
+		copy((*contentBytes)[pos:endPos], (*contentBytes)[match[8]:match[9]])
+
+		// Now copy what's left at the end of the string to the new end position
+		copy((*contentBytes)[endPos:], (*contentBytes)[match[9]:])
+		// now we reset the length
+
+		// our new section has length endPos - match[3]
+		// our old section has length match[9] - match[3]
+		(*contentBytes) = (*contentBytes)[:len((*contentBytes))-match[9]+endPos]
+		pos = endPos
+	}
+}
+
 // FindAllIssueReferences returns a list of unvalidated references found in a string.
 func FindAllIssueReferences(content string) []IssueReference {
 	// Need to convert fully qualified html references to local system to #/! short codes
 	contentBytes := []byte(content)
 	if re := getGiteaIssuePullPattern(); re != nil {
-		// We will iterate through the content, rewrite and simplify full references.
-		//
-		// We want to transform something like:
-		//
-		// 0         1         2         3         4         5         6
-		// 012345678901234567890123456789012345678901234567890123456789012345789
-		// this is a https://ourgitea.com/git/owner/repo/issues/123456789, foo
-		// https://ourgitea.com/git/owner/repo/pulls/123456789
-		//
-		// Into something like:
-		//
-		// this is a #123456789, foo
-		// !123456789
-
-		pos := 0
-		for {
-			// re looks for something like: (\s|^|\(|\[)https://ourgitea.com/git/(owner/repo)/(issues)/(123456789)(?:\s|$|\)|\]|[:;,.?!]\s|[:;,.?!]$)
-			match := re.FindSubmatchIndex(contentBytes[pos:])
-			if match == nil {
-				break
-			}
-			// match is a bunch of indices into the content from pos onwards so if our content is:
-
-			// To simplify things let's just add pos to all of the indices in match
-			for i := range match {
-				match[i] += pos
-			}
-
-			// match[0]-match[1] is whole string
-			// match[2]-match[3] is preamble
-
-			// move the position to the end of the preamble
-			pos = match[3]
-
-			// match[4]-match[5] is owner/repo
-			// now copy the owner/repo to end of the preamble
-			endPos := pos + match[5] - match[4]
-			copy(contentBytes[pos:endPos], contentBytes[match[4]:match[5]])
-
-			// move the current position to the end of the newly copied owner/repo
-			pos = endPos
-
-			// Now set the issue/pull marker:
-			//
-			// match[6]-match[7] == 'issues'
-			contentBytes[pos] = '#'
-			if string(contentBytes[match[6]:match[7]]) == "pulls" {
-				contentBytes[pos] = '!'
-			}
-			pos++
-
-			// Then add the issue/pull number
-			//
-			// match[8]-match[9] is the number
-			endPos = pos + match[9] - match[8]
-			copy(contentBytes[pos:endPos], contentBytes[match[8]:match[9]])
-
-			// Now copy what's left at the end of the string to the new end position
-			copy(contentBytes[endPos:], contentBytes[match[9]:])
-			// now we reset the length
-
-			// our new section has length endPos - match[3]
-			// our old section has length match[9] - match[3]
-			contentBytes = contentBytes[:len(contentBytes)-match[9]+endPos]
-			pos = endPos
-		}
+		convertFullHTMLReferencesToShortRefs(re, &contentBytes)
 	} else {
 		log.Debug("No GiteaIssuePullPattern pattern")
 	}

--- a/modules/references/references.go
+++ b/modules/references/references.go
@@ -240,30 +240,67 @@ func FindAllIssueReferences(content string) []IssueReference {
 	// Need to convert fully qualified html references to local system to #/! short codes
 	contentBytes := []byte(content)
 	if re := getGiteaIssuePullPattern(); re != nil {
+		// We will iterate through the content, rewrite and simplify full references.
+		//
+		// We want to transform something like:
+		//
+		// 0         1         2         3         4         5         6
+		// 012345678901234567890123456789012345678901234567890123456789012345789
+		// this is a https://ourgitea.com/git/owner/repo/issues/123456789, foo
+		// https://ourgitea.com/git/owner/repo/pulls/123456789
+		//
+		// Into something like:
+		//
+		// this is a #123456789, foo
+		// !123456789
+
 		pos := 0
 		for {
+			// re looks for something like: (\s|^|\(|\[)https://ourgitea.com/git/(owner/repo)/(issues)/(123456789)(?:\s|$|\)|\]|[:;,.?!]\s|[:;,.?!]$)
 			match := re.FindSubmatchIndex(contentBytes[pos:])
 			if match == nil {
 				break
 			}
+			// match is a bunch of indices into the content from pos onwards so if our content is:
+
+			// To simplify things let's just add pos to all of the indices in match
+			for i := range match {
+				match[i] += pos
+			}
+
 			// match[0]-match[1] is whole string
 			// match[2]-match[3] is preamble
-			pos += match[3]
+
+			// move the position to the end of the preamble
+			pos = match[3]
+
 			// match[4]-match[5] is owner/repo
+			// now copy the owner/repo to end of the preamble
 			endPos := pos + match[5] - match[4]
 			copy(contentBytes[pos:endPos], contentBytes[match[4]:match[5]])
+
+			// move the current position to the end of the newly copied owner/repo
 			pos = endPos
+
+			// Now set the issue/pull marker:
+			//
 			// match[6]-match[7] == 'issues'
 			contentBytes[pos] = '#'
 			if string(contentBytes[match[6]:match[7]]) == "pulls" {
 				contentBytes[pos] = '!'
 			}
 			pos++
+
+			// Then add the issue/pull number
+			//
 			// match[8]-match[9] is the number
 			endPos = pos + match[9] - match[8]
 			copy(contentBytes[pos:endPos], contentBytes[match[8]:match[9]])
+
+			// Now copy what's left at the end of the string to the new end position
 			copy(contentBytes[endPos:], contentBytes[match[9]:])
 			// now we reset the length
+
 			// our new section has length endPos - match[3]
 			// our old section has length match[9] - match[3]
 			contentBytes = contentBytes[:len(contentBytes)-match[9]+endPos]

--- a/modules/references/references.go
+++ b/modules/references/references.go
@@ -257,9 +257,8 @@ func convertFullHTMLReferencesToShortRefs(re *regexp.Regexp, contentBytes *[]byt
 		if match == nil {
 			break
 		}
-		// match is a bunch of indices into the content from pos onwards so if our content is:
-
-		// To simplify things let's just add pos to all of the indices in match
+		// match is a bunch of indices into the content from pos onwards so
+		// to simplify things let's just add pos to all of the indices in match
 		for i := range match {
 			match[i] += pos
 		}

--- a/modules/references/references.go
+++ b/modules/references/references.go
@@ -240,8 +240,6 @@ func convertFullHTMLReferencesToShortRefs(re *regexp.Regexp, contentBytes *[]byt
 	//
 	// We want to transform something like:
 	//
-	// 0         1         2         3         4         5         6
-	// 012345678901234567890123456789012345678901234567890123456789012345789
 	// this is a https://ourgitea.com/git/owner/repo/issues/123456789, foo
 	// https://ourgitea.com/git/owner/repo/pulls/123456789
 	//

--- a/modules/references/references_test.go
+++ b/modules/references/references_test.go
@@ -5,6 +5,7 @@
 package references
 
 import (
+	"regexp"
 	"testing"
 
 	"code.gitea.io/gitea/modules/setting"
@@ -27,6 +28,26 @@ type testResult struct {
 	RefLocation    *RefSpan
 	ActionLocation *RefSpan
 	TimeLog        string
+}
+
+func TestConvertFullHTMLReferencesToShortRefs(t *testing.T) {
+	re := regexp.MustCompile(`(\s|^|\(|\[)` +
+		regexp.QuoteMeta("https://ourgitea.com/git/") +
+		`([0-9a-zA-Z-_\.]+/[0-9a-zA-Z-_\.]+)/` +
+		`((?:issues)|(?:pulls))/([0-9]+)(?:\s|$|\)|\]|[:;,.?!]\s|[:;,.?!]$)`)
+	test := `this is a https://ourgitea.com/git/owner/repo/issues/123456789, foo
+https://ourgitea.com/git/owner/repo/pulls/123456789
+  And https://ourgitea.com/git/owner/repo/pulls/123
+`
+	expect := `this is a owner/repo#123456789, foo
+owner/repo!123456789
+  And owner/repo!123
+`
+
+	contentBytes := []byte(test)
+	convertFullHTMLReferencesToShortRefs(re, &contentBytes)
+	result := string(contentBytes)
+	assert.EqualValues(t, expect, result)
 }
 
 func TestFindAllIssueReferences(t *testing.T) {

--- a/modules/references/references_test.go
+++ b/modules/references/references_test.go
@@ -107,6 +107,13 @@ func TestFindAllIssueReferences(t *testing.T) {
 			},
 		},
 		{
+			"This http://gitea.com:3000/user4/repo5/pulls/202 yes. http://gitea.com:3000/user4/repo5/pulls/203 no",
+			[]testResult{
+				{202, "user4", "repo5", "202", true, XRefActionNone, nil, nil, ""},
+				{203, "user4", "repo5", "203", true, XRefActionNone, nil, nil, ""},
+			},
+		},
+		{
 			"This http://GiTeA.COM:3000/user4/repo6/pulls/205 yes.",
 			[]testResult{
 				{205, "user4", "repo6", "205", true, XRefActionNone, nil, nil, ""},

